File: client/goAPI/types/TGSystemObject.go
package types

import "bytes"

/**
 * Copyright 2018-19 TIBCO Software Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF DirectionAny KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * File name: TGSystemObject.go
 * Created on: Oct 13, 2018
 * Created by: achavan
 * SVN id: $id: $
 *
 */

// ======= System Types =======
type TGSystemType int

const (
	SystemTypeInvalid TGSystemType = -1
	SystemTypeEntity  TGSystemType = -2
)
const (
	SystemTypeAttributeDescriptor TGSystemType = iota
	SystemTypeNode
	SystemTypeEdge
	SystemTypeIndex
	SystemTypePrincipal
	SystemTypeRole
	SystemTypeSequence
	SystemTypeMaxSysObject
)

func (systemType TGSystemType) String() string {
	// Use a buffer for efficient string concatenation
	var buffer bytes.Buffer
	buffer.WriteString("")

	if systemType&SystemTypeInvalid == SystemTypeInvalid {
		buffer.WriteString("SystemTypeInvalid")
	} else if systemType&SystemTypeAttributeDescriptor == SystemTypeAttributeDescriptor {
		buffer.WriteString("SystemTypeAttributeDescriptor")
	} else if systemType&SystemTypeEntity == SystemTypeEntity {
		buffer.WriteString("SystemTypeEntity")
	} else if systemType&SystemTypeNode == SystemTypeNode {
		buffer.WriteString("SystemTypeNode")
	} else if systemType&SystemTypeEdge == SystemTypeEdge {
		buffer.WriteString("SystemTypeEdge")
	} else if systemType&SystemTypeIndex == SystemTypeIndex {
		buffer.WriteString("SystemTypeIndex")
	} else if systemType&SystemTypePrincipal == SystemTypePrincipal {
		buffer.WriteString("SystemTypePrincipal")
	} else if systemType&SystemTypeRole == SystemTypeRole {
		buffer.WriteString("SystemTypeRole")
	} else if systemType&SystemTypeSequence == SystemTypeSequence {
		buffer.WriteString("SystemTypeSequence")
	} else if systemType&SystemTypeMaxSysObject == SystemTypeMaxSysObject {
		buffer.WriteString("SystemTypeMaxSysObject")
	}
	if buffer.Len() == 0 {
		return ""
	}
	return buffer.String()
}

type TGSystemObject interface {
	TGSerializable
	// GetName gets the system object's name
	GetName() string
	// GetSystemType gets the system object's type
	GetSystemType() TGSystemType
}

File: api/java/src/com/tibco/tgdb/model/TGSystemObject.java

/**
 * Copyright 2019 TIBCO Software Inc. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except 
 * in compliance with the License.
 * A copy of the License is included in the distribution package with this file.
 * You also may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * File name : TGSystemObject.java
 * Created by: suresh
 * Created on: 04/16/2016
 * SVN Id: $Id: TGSystemObject.java 3143 2019-04-26 00:16:32Z nimish $
 */

package com.tibco.tgdb.model;

import com.tibco.tgdb.pdu.TGSerializable;


public interface TGSystemObject extends TGSerializable{

    enum TGSystemType {
    	InvalidType(-1),
    	AttributeDescriptor(0),
    	NodeType(1),
        EdgeType(2),
        Index(3),
        Principle(4),
        Role(5),
        Sequence(6),
        MaxSysObjectTypes(7);

        private final int type;

    	TGSystemType(int type) {
    		this.type = type;
    	}
    	
    	public int type() {
    		return type;
    	}

    	public static TGSystemType fromValue(int type) {
        	for (TGSystemType st : TGSystemType.values()) {
            	if (type == st.type) return st;
        	}
        	return InvalidType;
    	}
    }

    /**
     * Get the system desc enum
     * @return the system desc of the object
     */
    TGSystemType getSystemType();
    
    /**
     * Get the desc name.
     * @return the name of the object
     */
    String getName();

}
